### PR TITLE
Add unit tests for mappers

### PR DESCRIPTION
**Pull request by HristianIliev**

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -54,6 +54,18 @@
 				<artifactId>hamcrest</artifactId>
 				<version>${hamcrest.version}</version>
 			</dependency>
+      <dependency>
+			  <groupId>junit</groupId>
+			  <artifactId>junit</artifactId>
+			  <version>4.12</version>
+			  <scope>test</scope>
```
Comment (kuroidoruido) [inline pom.xml]: why adding junit 4 when there is always junit 5 (junit jupiter)?

File: pom.xml [hunk 2]
```diff
@@ -54,6 +54,18 @@
...
+		  </dependency>
+      <dependency>
+			  <groupId>org.mockito</groupId>
+			  <artifactId>mockito-all</artifactId>
+			  <version>1.10.19</version>
```
Comment (kuroidoruido) [inline pom.xml]: you can add mockito but if you add the latest version of it

File: pom.xml [hunk 3]
```diff
@@ -54,6 +54,18 @@
 				<artifactId>hamcrest</artifactId>
 				<version>${hamcrest.version}</version>
 			</dependency>
+      <dependency>
+			  <groupId>junit</groupId>
+			  <artifactId>junit</artifactId>
+			  <version>4.12</version>
```
Comment (kuroidoruido) [inline pom.xml]: you should follow the existing code style and put all dependency versions as properties

File: shielddb-gson-backend/src/test/java/nf/fr/k49/shielddb/gson/BottomShieldTest.java [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+
+    @Before
+    public void setUp() throws IOException {
+        String storagePath = BASE_DIR + "/User.json";
+
+        storage = spy(new FileStorage(storagePath));
+
+        String json = "[ \"String1\", \"String2\", \"String3\" ]";
+
+        storage.write(json);
```
Comment (kuroidoruido) [inline shielddb-gson-backend/src/test/java/nf/fr/k49/shielddb/gson/BottomShieldTest.java:33]: why did you add blank lines between every line of this method?

File: shielddb-gson-backend/src/test/java/nf/fr/k49/shielddb/gson/BottomShieldTest.java [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+
+        BottomShield<String> bottomShield = new BottomShield<>(new ShieldDBGson<>(), storage);
+
+        boolean saved = bottomShield.add("String4");
+
+        assertTrue(saved);
+
+        List<String> savedList = mapper.toList(storage.read());
+
+        assertThat(savedList.size(), is(4));
```
Comment (kuroidoruido) [inline shielddb-gson-backend/src/test/java/nf/fr/k49/shielddb/gson/BottomShieldTest.java:72]: why did you add blank lines between every line of this method?

File: shielddb-gson-backend/src/test/java/nf/fr/k49/shielddb/gson/ShieldDBGsonTest.java [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+        String json = "[ \"String1\", \"String2\", \"String3\" ]";
+
+        ShieldDBGson<String> shieldDBGson = new ShieldDBGson<>(new Gson(), GsonTypeUtils.getType());
+
+        List<String> list = shieldDBGson.toList(json);
+
+        assertThat(list.size(), is(3));
+        assertThat(list, containsInAnyOrder("String1", "String2", "String3"));
+    }
+
```
Comment (kuroidoruido) [inline shielddb-gson-backend/src/test/java/nf/fr/k49/shielddb/gson/ShieldDBGsonTest.java:43]: you did not test the toJson() method. Can you add tests on this method?

**Review by kuroidoruido — CHANGES_REQUESTED** [review]

File: shielddb-gson-backend/src/test/java/nf/fr/k49/shielddb/gson/ShieldDBGsonTest.java (hunk 6 above)
Comment (HristianIliev) [reply shielddb-gson-backend/src/test/java/nf/fr/k49/shielddb/gson/ShieldDBGsonTest.java:43]: The toJson() method contains one line of code "return gson.toJson(list);". As we can see this method only propagates the internal business logic of the gson external library. If I was to write a unit test for this method all that this would achieve is to test the gson library and not any productive code in ShieldDBGson. If you think that this would achieve anything I guess that I can add another unit test, but in my opinion it is not needed in this case.

File: shielddb-gson-backend/src/test/java/nf/fr/k49/shielddb/gson/BottomShieldTest.java (hunk 5 above)
Comment (HristianIliev) [reply shielddb-gson-backend/src/test/java/nf/fr/k49/shielddb/gson/BottomShieldTest.java:72]: If I were to remove these blank lines it would mean that the lines that do not have a blank line between them are logically identical, which they are not, hence the separation with blank lines. Also the blank lines increase the readability of the test. But if you want me to remove them, I guess that I do not have such a big problem with that.

File: shielddb-gson-backend/src/test/java/nf/fr/k49/shielddb/gson/BottomShieldTest.java (hunk 4 above)
Comment (HristianIliev) [reply shielddb-gson-backend/src/test/java/nf/fr/k49/shielddb/gson/BottomShieldTest.java:33]: If I were to remove these blank lines it would mean that the lines that do not have a blank line between them are logically identical, which they are not, hence the separation with blank lines. Also the blank lines increase the readability of the test. But if you want me to remove them, I guess that I do not have such a big problem with that.

File: pom.xml (hunk 2 above)
Comment (HristianIliev) [reply pom.xml]: This is the latest version of the mockito library. Mockito version 2.0.x is still in beta. Here are all dispalyed all of the versions of this library: https://mvnrepository.com/artifact/org.mockito/mockito-all

File: pom.xml (hunk 1 above)
Comment (HristianIliev) [reply pom.xml]: Applied

File: pom.xml (hunk 3 above)
Comment (HristianIliev) [reply pom.xml]: Applied

File: pom.xml (hunk 2 above)
Comment (kuroidoruido) [reply pom.xml]: But in another PR, the dependency who is imported is mickito-core 3.5.13, the official website indicate the latest version of mockito is 3.5.12: https://site.mockito.org/, the github has a release on 3.5.13 https://github.com/mockito/mockito. I think mockito-all is no more used, it's mockito-core we should use

File: shielddb-gson-backend/src/test/java/nf/fr/k49/shielddb/gson/BottomShieldTest.java (hunk 4 above)
Comment (kuroidoruido) [reply shielddb-gson-backend/src/test/java/nf/fr/k49/shielddb/gson/BottomShieldTest.java:33]: For me we should add a blank line between code portion that are logically distinct. In my mind I think if we add a blank line maybe we just need to create a function/method. In test case, I add a blank line between the 3 steps (given / when / then) of the test case but that all. Here in the setup you just make "storage initialization" so it's one thing to do, so yeah I think it's more readable to remove these blank lines

File: shielddb-gson-backend/src/test/java/nf/fr/k49/shielddb/gson/BottomShieldTest.java (hunk 5 above)
Comment (kuroidoruido) [reply shielddb-gson-backend/src/test/java/nf/fr/k49/shielddb/gson/BottomShieldTest.java:72]: there is too much blank line, you should isolate given / when / then or given / (when/then) / (when/then) like that:

```java
        ShieldDBGson<String> mapper = new ShieldDBGson<>();
        BottomShield<String> bottomShield = new BottomShield<>(new ShieldDBGson<>(), storage);

        boolean saved = bottomShield.add("String4");
        assertTrue(saved);

        List<String> savedList = mapper.toList(storage.read());
        assertThat(savedList.size(), is(4));
```

This way our brain already know there is 3 steps: initialization, testing bottomShield.add() works, testing mapper.toList() works

File: shielddb-gson-backend/src/test/java/nf/fr/k49/shielddb/gson/ShieldDBGsonTest.java (hunk 6 above)
Comment (kuroidoruido) [reply shielddb-gson-backend/src/test/java/nf/fr/k49/shielddb/gson/ShieldDBGsonTest.java:43]: First point: today it contains only one line of code, but in the future it may not.
Second point: best unit tests are black boxed, so you should write tests as you did not read the code.
Third point: test are for me a living documentation, so do not testing a method is like they do nothing, so testing ShieldDBGson involve you to test every public method, this way anywhere we will use this method, if we change toJson(), tests will tell us what was expected.
